WinError 10013: "An attempt was made to access a socket in a way forbidden by its access permissions"

Since this week, the weekly all-test run fails all indication unit tests on Windows/Python3.7/latest with this error
(from the log of https://github.com/pywbem/pywbem/runs/5526357627?check_suite_focus=true).
The error does not happen on any other environment, i.e. not on Windows/Python3.7/minimum, and not on Windows with other Python versions between 3.5 and 3.10 (TODO(Andy): Verify)

```
________________________ test_WBEMListener_start_stop _________________________

    def test_WBEMListener_start_stop():
        """
        Test starting and stopping of the listener.
        """
    
        host = 'localhost'
        http_port = '50000'
    
        listener = WBEMListener(host, http_port)
        assert listener.http_started is False
        assert listener.https_started is False
    
>       listener.start()

tests\unittest\pywbem\test_indicationlistener.py:439: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pywbem\_listener.py:985: in start
    ListenerRequestHandler)
C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\socketserver.py:452: in __init__
    self.server_bind()
C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\http\server.py:137: in server_bind
    socketserver.TCPServer.server_bind(self)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pywbem._listener.ThreadedHTTPServer object at 0x0000029509897248>

    def server_bind(self):
        """Called by constructor to bind the socket.
    
        May be overridden.
    
        """
        if self.allow_reuse_address:
            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
>       self.socket.bind(self.server_address)
E       OSError: [WinError 10013] An attempt was made to access a socket in a way forbidden by its access permissions

C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\socketserver.py:466: OSError
```

--- a/changes/2862.fix.rst
+++ b/changes/2862.fix.rst
@@ -1,0 +1,5 @@
+Test: Fixed the occasional permission errors during indication listener tests
+that occurred on Windows, by letting the operating system assign the port for
+the listener under test, instead of hard coding them.
+This required adding new properties 'http_bound_port' and 'https_bound_port'
+to the :class:`WBEMListener` class that indicate the assigned ports.

--- a/pywbem/_listener.py
+++ b/pywbem/_listener.py
@@ -998,16 +998,24 @@ class WBEMListener:
             IP addressing family.
 
           http_port (:class:`py:str` or :class:`py:int`):
-            HTTP port at which this listener can be reached. At
-            least one port (HTTP or HTTPS) must be set. Both the http and
-            https ports can be set.
+            HTTP port to be used for the listener.
+            A value of 0 or '0' means that the operating system selects the
+            port.
 
             `None` means not to set up a port for HTTP.
 
+            At least one port (HTTP or HTTPS) must be set up. Both the HTTP and
+            HTTPS ports can be set up.
+
           https_port (:class:`py:str` or :class:`py:int`):
-            HTTPS port at which this listener can be reached.
+            HTTPS port to be used for the listener.
+            A value of 0 or '0' means that the operating system selects the
+            port.
 
             `None` means not to set up a port for HTTPS.
+
+            At least one port (HTTP or HTTPS) must be set up. Both the HTTP and
+            HTTPS ports can be set up.
 
           certfile (str):
             File path of certificate file to be used as server certificate
@@ -1183,7 +1191,11 @@ class WBEMListener:
     @property
     def http_port(self):
         """
-        int: HTTP port at which this listener can be reached.
+        int: HTTP port to be used for the listener.
+
+        A value of 0 means that the operating system should assign the port.
+        When the listener is running, the port that is used by the listener is
+        available in :attr:`http_bound_port`.
 
         `None` means there is no port set up for HTTP.
         """
@@ -1192,16 +1204,46 @@ class WBEMListener:
     @property
     def https_port(self):
         """
-        int: HTTPS port at which this listener can be reached.
+        int: HTTPS port to be used for the listener.
+
+        A value of 0 means that the operating system should assign the port.
+        When the listener is running, the port that is used by the listener is
+        available in :attr:`https_bound_port`.
 
         `None` means there is no port set up for HTTPS.
         """
         return self._https_port
 
     @property
+    def http_bound_port(self):
+        """
+        int: HTTP port this listener is bound to when started.
+
+        `None` when the listener is not running.
+
+        *New in pywbem 1.10.*
+        """
+        if self._http_server is not None:
+            return getattr(self._http_server, "server_port", None)
+        return None
+
+    @property
+    def https_bound_port(self):
+        """
+        int: HTTPS port this listener is bound to when started.
+
+        `None` when the listener is not running.
+
+        *New in pywbem 1.10.*
+        """
+        if self._https_server is not None:
+            return getattr(self._https_server, "server_port", None)
+        return None
+
+    @property
     def http_started(self):
         """
-        bool: Boolean indicating whether the listener is started for the HTTP
+        bool: Boolean indicating whether the listener is running for the HTTP
         port.
 
         If no port is set up for HTTP, `False` is returned.
@@ -1213,7 +1255,7 @@ class WBEMListener:
     @property
     def https_started(self):
         """
-        bool: Boolean indicating whether the listener is started for the HTTPS
+        bool: Boolean indicating whether the listener is running for the HTTPS
         port.
 
         If no port is set up for HTTPS, `False` is returned.
@@ -1381,7 +1423,7 @@ class WBEMListener:
         self._callback_thread.start()
 
         try:
-            if self._http_port:
+            if self._http_port is not None:
                 if not self._http_server:
                     self.logger.info(
                         "Creating threaded HTTP server for host %r on port %s",
@@ -1421,7 +1463,7 @@ class WBEMListener:
                 self._http_server = None
                 self._http_thread = None
 
-            if self._https_port:
+            if self._https_port is not None:
                 if not self._https_server:
 
                     self.logger.info(

--- a/tests/end2endtest/test_indications.py
+++ b/tests/end2endtest/test_indications.py
@@ -55,7 +55,7 @@ def test_indications(wbem_connection):  # noqa: F811
     # Use the empty string host that allows receving indications addressed to
     # any valid valid destination on the system.
     listener_host = ""
-    port = 5000
+    port = 0  # Let the OS assign the port
     log = True
     verbose = True
 

--- a/tests/unittest/pywbem/test_indicationlistener.py
+++ b/tests/unittest/pywbem/test_indicationlistener.py
@@ -502,21 +502,33 @@ def test_WBEMListener_start_stop():
     """
 
     host = 'localhost'
-    http_port = '50000'  # Intentionally string-typed
+    http_port = '0'  # Intentionally string-typed, 0 = let OS assign the port
 
     listener = WBEMListener(host, http_port)
     assert listener.http_started is False
+    assert listener.http_port == int(http_port)
+    assert listener.http_bound_port is None
     assert listener.https_started is False
+    assert listener.https_port is None
+    assert listener.https_bound_port is None
 
     try:
         listener.start()
         assert listener.http_started is True
+        assert listener.http_port == int(http_port)
+        assert listener.http_bound_port is not None
         assert listener.https_started is False
+        assert listener.https_port is None
+        assert listener.https_bound_port is None
 
     finally:
         listener.stop()
         assert listener.http_started is False
+        assert listener.http_port == int(http_port)
+        assert listener.http_bound_port is None
         assert listener.https_started is False
+        assert listener.https_port is None
+        assert listener.https_bound_port is None
 
 
 @log_entry_exit
@@ -526,19 +538,19 @@ def test_WBEMListener_port_in_use():
     """
 
     host = 'localhost'
-
-    # Don't use this port in other tests, to be on the safe side
-    # as far as port reuse is concerned.
-    http_port = '50001'  # Intentionally string-typed
+    http_port = 0  # Let the OS assign the port
 
     exp_exc_type = ListenerPortError
 
-    listener1 = WBEMListener(host, http_port)
-    listener2 = WBEMListener(host, http_port)
+    listener1 = WBEMListener(host, http_port)  # Let the OS assign the port
+    listener2 = None
 
     try:
         listener1.start()
         assert listener1.http_started is True
+
+        # Set up listener 2 with the same port as listener 1
+        listener2 = WBEMListener(host, listener1.http_bound_port)
 
         try:
             # The code to be tested
@@ -557,7 +569,8 @@ def test_WBEMListener_port_in_use():
 
     finally:
         listener1.stop()
-        listener2.stop()
+        if listener2:
+            listener2.stop()
 
 
 @log_entry_exit
@@ -568,9 +581,8 @@ def test_WBEMListener_context_mgr():
 
     host = 'localhost'
 
-    # Don't use this port in other tests, to be on the safe side
-    # as far as port reuse is concerned.
-    http_port = '50002'  # Intentionally string-typed
+    http_port = 0  # Let the OS assign the port
+    http_bound_port = None
 
     # The code to be tested (is the context manager)
     with WBEMListener(host, http_port) as listener1:
@@ -580,12 +592,13 @@ def test_WBEMListener_context_mgr():
 
         listener1.start()
         assert listener1.http_started is True
+        http_bound_port = listener1.http_bound_port
 
     # Verify that CM exit stops the listener
     assert listener1.http_started is False
 
     # Verify that the TCP/IP port can be used again
-    listener2 = WBEMListener(host, http_port)
+    listener2 = WBEMListener(host, http_bound_port)
     try:
         listener2.start()
         assert listener2.http_started is True
@@ -836,7 +849,7 @@ def test_WBEMListener_send_indications(
         host = '127.0.0.1'
     else:
         host = 'localhost'
-    http_port = 50000
+    http_port = 0  # Let the OS assign the port
 
     listener = WBEMListener(host, http_port, max_ind_queue_size=max_queue)
     listener.add_callback(process_indication_callback)
@@ -846,8 +859,9 @@ def test_WBEMListener_send_indications(
 
     try:
         listener.start()
+        http_bound_port = listener.http_bound_port
         start_time = time()
-        url = f'http://{host}:{http_port}'
+        url = f'http://{host}:{http_bound_port}'
         cim_protocol_version = '1.4'
         headers = {
             'Content-Type': 'application/xml; charset=utf-8',
@@ -1004,8 +1018,7 @@ def test_WBEMListener_incorrect_method(method, exp_status):
     """
 
     host = 'localhost'
-    http_port = 50000
-    url = f'http://{host}:{http_port}'
+    http_port = 0  # Let the OS assign the port
     headers = {
         'Content-Type': 'application/xml; charset=utf-8',
         'CIMExport': 'MethodRequest',
@@ -1019,6 +1032,9 @@ def test_WBEMListener_incorrect_method(method, exp_status):
     try:
         listener.add_callback(process_indication_callback)
         listener.start()
+
+        http_bound_port = listener.http_bound_port
+        url = f'http://{host}:{http_bound_port}'
 
         # The code to be tested is running in listener thread
         response = requests.request(method, url, headers=headers, timeout=4)
@@ -1127,8 +1143,7 @@ def test_WBEMListener_incorrect_headers(desc, headers, exp_status, exp_headers):
     """
 
     host = 'localhost'
-    http_port = 50000
-    url = f'http://{host}:{http_port}'
+    http_port = 0  # Let the OS assign the port
     # headers = copy(headers)
 
     listener = WBEMListener(host, http_port)
@@ -1136,6 +1151,9 @@ def test_WBEMListener_incorrect_headers(desc, headers, exp_status, exp_headers):
     try:
         listener.add_callback(process_indication_callback)
         listener.start()
+
+        http_bound_port = listener.http_bound_port
+        url = f'http://{host}:{http_bound_port}'
 
         # The code to be tested is running in listener thread
         response = post_bsl(url, headers=headers, data=None)
@@ -1282,8 +1300,7 @@ def test_WBEMListener_incorrect_payload1(
     """
 
     host = 'localhost'
-    http_port = 50000
-    url = f'http://{host}:{http_port}'
+    http_port = 0  # Let the OS assign the port
     headers = {
         'Content-Type': 'application/xml; charset=utf-8',
         'CIMExport': 'MethodRequest',
@@ -1297,6 +1314,9 @@ def test_WBEMListener_incorrect_payload1(
     try:
         listener.add_callback(process_indication_callback)
         listener.start()
+
+        http_bound_port = listener.http_bound_port
+        url = f'http://{host}:{http_bound_port}'
 
         # The code to be tested is running in listener thread
         response = post_bsl(url, headers=headers, data=payload)
@@ -1448,8 +1468,7 @@ def test_WBEMListener_incorrect_payload2(
     """
 
     host = 'localhost'
-    http_port = 50000
-    url = f'http://{host}:{http_port}'
+    http_port = 0  # Let the OS assign the port
     headers = {
         'Content-Type': 'application/xml; charset=utf-8',
         'CIMExport': 'MethodRequest',
@@ -1463,6 +1482,9 @@ def test_WBEMListener_incorrect_payload2(
     try:
         listener.add_callback(process_indication_callback)
         listener.start()
+
+        http_bound_port = listener.http_bound_port
+        url = f'http://{host}:{http_bound_port}'
 
         # The code to be tested is running in listener thread
         response = post_bsl(url, headers=headers, data=payload)

